flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/JoinUtil.scala

```diff
@@ -0,0 +1,468 @@
+  }
+
+  /**
+    * Generate other non-equi condition function
+    * @param  config   table env config
+    * @param  joinType  join type to determain whether input can be null
+    * @param  leftType  left stream type
+    * @param  rightType  right stream type
+    * @param  returnType   return type
+    * @param  otherCondition   non-equi condition
+    * @param  ruleDescription  rule description
+    */
+  private[flink] def generateJoinFunction(
+    config: TableConfig,
+    joinType: JoinRelType,
+    leftType: TypeInformation[Row],
+    rightType: TypeInformation[Row],
+    returnType: RowSchema,
+    otherCondition: RexNode,
+    ruleDescription: String) = {
+
+    // whether input can be null
+    val nullCheck = joinType match {
+      case JoinRelType.INNER => false
+      case JoinRelType.LEFT  => true
+      case JoinRelType.RIGHT => true
+      case JoinRelType.FULL  => true
+    }
+
+    // generate other non-equi function code
+    val generator = new CodeGenerator(
+      config,
+      nullCheck,
+      leftType,
+      Some(rightType))
+
+    val conversion = generator.generateConverterResultExpression(
+      returnType.physicalTypeInfo,
+      returnType.physicalType.getFieldNames)
+
+    // if other condition is literal(true), then output the result directly
+    val body = if (otherCondition.isAlwaysTrue) {
+      s"""
+         |${conversion.code}
+         |${generator.collectorTerm}.collect(${conversion.resultTerm});
+         |""".stripMargin
+    }
+    else {
+      val condition = generator.generateExpression(otherCondition)
+      s"""
+         |${condition.code}
+         |if (${condition.resultTerm}) {
+         |  ${conversion.code}
+         |  ${generator.collectorTerm}.collect(${conversion.resultTerm});
+         |}
+         |""".stripMargin
+    }
+
+    val genFunction = generator.generateFunction(
+      ruleDescription,
+      classOf[FlatJoinFunction[Row, Row, Row]],
+      body,
+      returnType.physicalTypeInfo)
+
+    genFunction
+  }
+
+  private case class TimeSingleCondition(
+      timeType: RelDataType,
+      leftExpr: RexNode,
+      rightExpr: RexNode,
+      isEqual: Boolean,
+      originCall: RexNode)
+
+  val COMPARISON: util.Set[SqlKind] = EnumSet.of(
+    SqlKind.LESS_THAN,
+    SqlKind.GREATER_THAN,
+    SqlKind.GREATER_THAN_OR_EQUAL,
+    SqlKind.LESS_THAN_OR_EQUAL)
+
+  val EQUI_COMPARISON: util.Set[SqlKind] = EnumSet.of(
+    SqlKind.GREATER_THAN_OR_EQUAL,
+    SqlKind.LESS_THAN_OR_EQUAL)
+
+  val LESS_COMPARISON: util.Set[SqlKind] = EnumSet.of(
+    SqlKind.LESS_THAN,
+    SqlKind.LESS_THAN_OR_EQUAL)
+
+  val GREAT_COMPARISON: util.Set[SqlKind] = EnumSet.of(
+    SqlKind.GREATER_THAN,
+    SqlKind.GREATER_THAN_OR_EQUAL)
+
+  /**
+    * Analyze time-conditon to divide all time-condition into great and less condition
+    */
+  private def analyzeTimeCondition(
+    condition: RexNode,
+    greatCondition: util.List[TimeSingleCondition],
```

Comment:
If I understand the function correctly, it will not detect conditions like `t1.rowtime - t2.rowtime < 10.minutes`, right?